## Primeira análise de dados sobre astronomia

Vamos analisar a base de dados Planetary Systems com intuito de aprender a realizar uma boa analise de dados.

O método utilizado:

1. Análisar a base de dados (ver suas colunas)
2. Descartar colunas que não ajudam na análise (Ex. colunas de ID)
3. Tratar os dados (dados duplicados, nulos, tipos)
4. Análise inicial
5. Análise aprofundada
6. Apresentar os resultados

Dataset URL = https://exoplanetarchive.ipac.caltech.edu/cgi-bin/TblView/nph-tblView?app=ExoTbls&config=PS  

In [21]:
import pandas as pd

# ler a base de dados. 
df = pd.read_csv("exoplanet_dataset.csv", sep=",", comment="#", dtype={4: str , 5: str})

display(df)

,rowid,pl_name,hostname,pl_letter,hd_name,hip_name,tic_id,gaia_dr2_id,gaia_dr3_id,default_flag,...,rowupdate,pl_pubdate,releasedate,pl_nnotes,st_nphot,st_nrvc,st_nspec,pl_nespec,pl_ntranspec,pl_ndispec
0,1,11 Com b,11 Com,b,HD 107383,HIP 60202,TIC 72437047,Gaia DR2 3946945413106333696,Gaia DR3 3946945413106333696,1,...,2023-09-19,2023-08,2023-09-19,2.0,1.0,2.0,0.0,0.0,0.0,0.0
1,2,11 Com b,11 Com,b,HD 107383,HIP 60202,TIC 72437047,Gaia DR2 3946945413106333696,Gaia DR3 3946945413106333696,0,...,2014-05-14,2008-01,2014-05-14,2.0,1.0,2.0,0.0,0.0,0.0,0.0
2,3,11 Com b,11 Com,b,HD 107383,HIP 60202,TIC 72437047,Gaia DR2 3946945413106333696,Gaia DR3 3946945413106333696,0,...,2014-07-23,2011-08,2014-07-23,2.0,1.0,2.0,0.0,0.0,0.0,0.0
3,4,11 UMi b,11 UMi,b,HD 136726,HIP 74793,TIC 230061010,Gaia DR2 1696798367260229376,Gaia DR3 1696798367260229376,0,...,2018-04-25,2009-10,2014-05-14,0.0,1.0,1.0,0.0,0.0,0.0,0.0
4,5,11 UMi b,11 UMi,b,HD 136726,HIP 74793,TIC 230061010,Gaia DR2 1696798367260229376,Gaia DR3 1696798367260229376,1,...,2018-09-04,2017-03,2018-09-06,0.0,1.0,1.0,0.0,0.0,0.0,0.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
39471,39472,ups And d,ups And,d,HD 9826,HIP 7513,TIC 189576919,Gaia DR2 348020448377061376,Gaia DR3 348020482735930112,0,...,2021-09-20,2021-07,2021-09-20,5.0,1.0,10.0,1.0,0.0,0.0,0.0
39472,39473,ups Leo b,ups Leo,b,NaN,NaN,TIC 49430557,Gaia DR2 3794167001116433152,Gaia DR3 3794167001116704000,1,...,2022-01-10,2021-12,2022-01-10,0.0,0.0,0.0,0.0,0.0,0.0,0.0
39473,39474,xi Aql b,xi Aql,b,HD 188310,HIP 97938,TIC 375464367,Gaia DR2 4298361114750843904,Gaia DR3 4298361114753012992,0,...,2014-07-23,2011-08,2014-07-23,1.0,1.0,1.0,0.0,0.0,0.0,0.0
39474,39475,xi Aql b,xi Aql,b,HD 188310,HIP 97938,TIC 375464367,Gaia DR2 4298361114750843904,Gaia DR3 4298361114753012992,0,...,2014-05-14,2008-06,2014-05-14,1.0,1.0,1.0,0.0,0.0,0.0,0.0


In [22]:
# selecionar algumas colunas para trabalhar
colunas = [
    "pl_name", "hostname", "discoverymethod", 
    "disc_year", "pl_orbper", "pl_rade",
    "st_mass", "st_rad", "st_teff", "sy_dist" 
]

df = df[colunas]

## Colunas selecionadas para a análise:

- (A massa do planeta não foi utilizada por cerca de 90% das linhas estarem constando NaN)

- pl_name -> nome do exoplaneta
- hostname -> nome da estrela
- discoverymethod -> método utilizado para descobrir o exoplaneta
- disc_year -> ano que o exoplaneta foi descoberto
- pl_orbper -> período orbital do exoplaneta (em dias)
- pl_rade -> raio do exoplaneta (considerando o raio da Terra. Ex. pl_rade = 1.0 significa que é o mesmo raio da Terra)
- st_mass -> massa da estrela (considera a massa solar)
- st_rad -> raio da estrela (considera o raio do Sol)
- st_teff -> temperatura da superfície da estrela (em Kelvin - K)
- sy_dist -> distância até a Terra (em persecs - pc, corresponde a ~3,26 anos luz)


In [23]:
# tratar os dados (null, NaN, tipos errados)

display(df.isna().sum())

# remove valores nulos (quando não for em uma quantidade muito grande)
df = df.dropna()
display(df.info())

# guardar dados do tipo texto em uma variável separada
nomes_sem_duplicatas = df.drop_duplicates("pl_name")
colunas_nao_numericas = nomes_sem_duplicatas[["pl_name", "hostname", "discoverymethod"]]

# fazer a média dos valores numerícos (Ex. "pl_rade") agrupados pelo nome do planeta
df = df.groupby("pl_name").mean(numeric_only=True)

# unir ambas as variáveis (colunas_nao_numericas e df) para formar a tabela com os dados completos 
df = colunas_nao_numericas.merge(df, on="pl_name")
display(df)

pl_name                0
hostname               0
discoverymethod        0
disc_year              2
pl_orbper           3373
pl_rade            12319
st_mass             6228
st_rad              3380
st_teff             3577
sy_dist              904
dtype: int64

<class 'pandas.core.frame.DataFrame'>
Index: 20943 entries, 116 to 39422
Data columns (total 10 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   pl_name          20943 non-null  object 
 1   hostname         20943 non-null  object 
 2   discoverymethod  20943 non-null  object 
 3   disc_year        20943 non-null  float64
 4   pl_orbper        20943 non-null  float64
 5   pl_rade          20943 non-null  float64
 6   st_mass          20943 non-null  float64
 7   st_rad           20943 non-null  float64
 8   st_teff          20943 non-null  float64
 9   sy_dist          20943 non-null  float64
dtypes: float64(7), object(3)
memory usage: 1.8+ MB


None

,pl_name,hostname,discoverymethod,disc_year,pl_orbper,pl_rade,st_mass,st_rad,st_teff,sy_dist
0,55 Cnc e,55 Cnc,Radial Velocity,2004.0,0.736620,1.965167,0.898000,0.951672,5225.000000,12.5855
1,AU Mic b,AU Mic,Transit,2020.0,8.463184,4.255036,0.536250,0.776500,3654.500000,9.7221
2,AU Mic c,AU Mic,Transit,2021.0,18.859244,2.850674,0.548333,0.785333,3639.333333,9.7221
3,BD+20 594 b,BD+20 594,Transit,2016.0,41.686548,2.387841,1.198287,0.988041,5745.000000,179.4610
4,BD-14 3065 b,BD-14 3065 A,Transit,2024.0,4.288973,21.590000,1.410000,2.350000,6935.000000,589.4230
...,...,...,...,...,...,...,...,...,...,...
4429,XO-6 b,XO-6,Transit,2016.0,3.765001,23.202630,1.470000,1.930000,6720.000000,235.4790
4430,XO-7 b,XO-7,Transit,2019.0,2.864142,15.373130,1.405000,1.480000,6250.000000,234.1490
4431,ZTF J1230-2655 b,ZTF J1230-2655,Transit,2025.0,0.235978,13.787046,0.646000,0.012300,10000.000000,186.6900
4432,ZTF J1828+2308 b,ZTF J1828+2308,Transit,2025.0,0.112007,11.130518,0.610000,0.013100,15900.000000,201.8270


In [24]:
import plotly.express as px

# verificar a porcentagem de cada método utilizado para descobrimento de explanetas
display(df["discoverymethod"].value_counts(normalize=True).map("{:.2%}".format))

# construir um gráfico que demonstre a quantidade de vezes que cada método para descobrimento de exoplanetas foi utilizado
grafico = px.histogram(
    df, 
    x="discoverymethod", 
    title="Métodos de Descobrir Exoplanetas", 
    subtitle="Percentual de métodos de descobrir exoplanetas",
    text_auto=True,
)
grafico.show()

discoverymethod
Transit                          98.89%
Radial Velocity                   0.70%
Transit Timing Variations         0.29%
Orbital Brightness Modulation     0.07%
Imaging                           0.05%
Name: proportion, dtype: object

In [25]:
# verificar a quantidade de exoplanetas que foram descobertos por ano
display(df["disc_year"].value_counts())

# construir um gráfico demonstrando a quantidade de exoplanetas que foram descobertos por ano e quais métodos foram utilizados
grafico = px.histogram(df, x="disc_year", color="discoverymethod")
grafico.show()

disc_year
2016.0    1388
2014.0     788
2021.0     445
2018.0     243
2023.0     223
2022.0     191
2024.0     185
2020.0     161
2025.0     143
2015.0     101
2019.0      96
2012.0      93
2013.0      81
2017.0      81
2011.0      80
2010.0      48
2026.0      20
2009.0      18
2007.0      17
2008.0      16
2004.0       7
2005.0       4
2006.0       3
1999.0       1
2001.0       1
Name: count, dtype: int64

In [39]:
# construindo um gráfico para mostrar estrelas filtradas
# filtro pela quantidade de massa da estrela no intervalo: 0.9 - 1.1 vezes a massa do Sol e com raio até duas vezes o raio do Sol
grafico = px.scatter(
    df[(df["st_mass"] > 0.9) & (df["st_mass"] < 1.1) & (df["st_rad"] <= 2)], 
    x="st_mass", 
    y="st_rad", 
    color="hostname",
    title="Estrelas similares ao Sol",
)
grafico.show()

# display para informar a quantidade de itens encontrados de acordo com o filtro aplicado no gráfico acima e outras informações como tipo de dado
display(df[(df["st_mass"] > 0.9) & (df["st_mass"] < 1.1) & (df["st_rad"] <= 2)].info())

<class 'pandas.core.frame.DataFrame'>
Index: 1698 entries, 6 to 4433
Data columns (total 10 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   pl_name          1698 non-null   object 
 1   hostname         1698 non-null   object 
 2   discoverymethod  1698 non-null   object 
 3   disc_year        1698 non-null   float64
 4   pl_orbper        1698 non-null   float64
 5   pl_rade          1698 non-null   float64
 6   st_mass          1698 non-null   float64
 7   st_rad           1698 non-null   float64
 8   st_teff          1698 non-null   float64
 9   sy_dist          1698 non-null   float64
dtypes: float64(7), object(3)
memory usage: 145.9+ KB


None

In [27]:
# varivel para armazenar linhas com estrelas semelhantes a algumas características físicas do Sol como massa e raio
sol_like = df[(df["st_mass"] > 0.9) & (df["st_mass"] < 1.1) & (df["st_rad"] <= 2)]

# filtrando as linhas restantes para armazenar aquelas que o planeta possui seu período orbital similar ao da Terra (até ~17% menor ou até ~91% maior) 
sol_like = sol_like[(sol_like["pl_orbper"] > 300) & (sol_like["pl_orbper"] < 700)]

# display de informações sobre os planetas encontrados de acordo com o filtro aplicado acima
display(sol_like.info())

# tabela organizando os planetas de acordo com seu período orbital (em dias) na ordem decrescente
display(sol_like.sort_values(by="pl_orbper", ascending=False))

<class 'pandas.core.frame.DataFrame'>
Index: 12 entries, 845 to 3367
Data columns (total 10 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   pl_name          12 non-null     object 
 1   hostname         12 non-null     object 
 2   discoverymethod  12 non-null     object 
 3   disc_year        12 non-null     float64
 4   pl_orbper        12 non-null     float64
 5   pl_rade          12 non-null     float64
 6   st_mass          12 non-null     float64
 7   st_rad           12 non-null     float64
 8   st_teff          12 non-null     float64
 9   sy_dist          12 non-null     float64
dtypes: float64(7), object(3)
memory usage: 1.0+ KB


None

,pl_name,hostname,discoverymethod,disc_year,pl_orbper,pl_rade,st_mass,st_rad,st_teff,sy_dist
1555,Kepler-150 f,Kepler-150,Transit,2017.0,637.209300,3.640000,0.970000,0.905339,5528.000000,891.092
3367,Kepler-87 c,Kepler-87,Transit,2013.0,464.604550,5.368571,1.009429,1.435429,5791.714286,1232.900
845,KIC 5437945 b,KIC 5437945,Transit,2015.0,440.785192,6.967571,1.077000,1.420000,6271.428571,1274.010
1728,Kepler-1636 b,Kepler-1636,Transit,2016.0,425.478988,3.352600,1.089600,1.065000,5890.800000,1947.540
1724,Kepler-1632 b,Kepler-1632,Transit,2016.0,392.265399,2.050000,0.906500,1.025000,6087.250000,716.760
2825,Kepler-452 b,Kepler-452,Transit,2015.0,384.844150,1.337500,0.957250,0.940500,5701.500000,551.727
1726,Kepler-1634 b,Kepler-1634,Transit,2016.0,374.876712,4.325000,1.057250,1.238000,5595.500000,617.339
870,KOI-351 h,KOI-351,Transit,2013.0,331.605267,10.670343,1.084222,1.140526,6135.570000,848.254
2452,Kepler-289 e,Kepler-289,Transit,2025.0,330.071558,2.940000,0.998000,1.115000,5933.000000,704.438
2959,Kepler-553 c,Kepler-553,Transit,2016.0,328.240043,11.411485,0.908750,0.874250,5348.500000,728.525


In [40]:
# função para inserir uma nova coluna chamada pl_class
def classificar_exoplanetas(pl_rade: float) -> str:
    if (pl_rade <= 0.8):
        return "mini_earth"
    if (pl_rade <= 1.25):
        return "earth-sized"
    if (pl_rade <= 1.75):
        return "super_earth"
    if (pl_rade <= 3):
        return "mini_neptune"
    if (pl_rade <= 6):
        return "neptune"
    else:
        return "gas_giant"
        
# receber valores para a nova coluna
sol_like["pl_class"] = sol_like["pl_rade"].apply(classificar_exoplanetas)

# display da tabela atualizada
display(sol_like)

# este gráfico mostra a distribuição dos planetas de acordo com seu raio
grafico = px.bar(
    sol_like.sort_values(by="pl_rade", ascending=False), 
    x="pl_rade", 
    y="pl_name", 
    color="pl_class",
    title="Exoplanetas de acordo com o raio",
    subtitle="Classificação dos exoplanetas considerando o raio"
)
grafico.show()

,pl_name,hostname,discoverymethod,disc_year,pl_orbper,pl_rade,st_mass,st_rad,st_teff,sy_dist,pl_class
845,KIC 5437945 b,KIC 5437945,Transit,2015.0,440.785192,6.967571,1.077000,1.420000,6271.428571,1274.010,gas_giant
870,KOI-351 h,KOI-351,Transit,2013.0,331.605267,10.670343,1.084222,1.140526,6135.570000,848.254,gas_giant
1555,Kepler-150 f,Kepler-150,Transit,2017.0,637.209300,3.640000,0.970000,0.905339,5528.000000,891.092,neptune
1724,Kepler-1632 b,Kepler-1632,Transit,2016.0,392.265399,2.050000,0.906500,1.025000,6087.250000,716.760,mini_neptune
1726,Kepler-1634 b,Kepler-1634,Transit,2016.0,374.876712,4.325000,1.057250,1.238000,5595.500000,617.339,neptune
1728,Kepler-1636 b,Kepler-1636,Transit,2016.0,425.478988,3.352600,1.089600,1.065000,5890.800000,1947.540,neptune
1867,Kepler-1746 b,Kepler-1746,Transit,2021.0,321.212624,6.664435,0.999571,1.136714,6050.428571,979.541,gas_giant
2452,Kepler-289 e,Kepler-289,Transit,2025.0,330.071558,2.940000,0.998000,1.115000,5933.000000,704.438,mini_neptune
2825,Kepler-452 b,Kepler-452,Transit,2015.0,384.844150,1.337500,0.957250,0.940500,5701.500000,551.727,super_earth
2844,Kepler-47 c,Kepler-47,Transit,2012.0,303.137000,4.620000,1.049000,0.963000,5636.000000,1025.020,neptune


## Classificação dos exoplanetas de acordo com seu raio:

- Mini Terra (pl_rade > 0.5 e <= 0.8)

- Terra (pl_rade > 0.8 e <= 1.25)

- Super Terra (pl_rade > 1.25 e <= 1.75)

- Mini Netuno (pl_rade > 1.75 e <= 3)

- Netuno (pl_rade > 3 e <= 6)

- Gigante Gasoso (pl_rade > 6 e <= ~14.3)

(para informações a respeito do tipo de cada um desses planetas acesse o link do site da Nasa abaixo)

Fonte : https://science.nasa.gov/exoplanets/planet-types/#h-a-galaxy-of-stars-and-planets

## Conclusões

1698 estrelas similares ao Sol foram encontradas;

Essas estrelas foram encontradas através da seguinte filtragem:
 - A estrela candidata deve ter massa (st_mass) entre 0.9 e 1.1 (em massa solar, ou seja, 1 = 1 vez a massa do Sol);
 - A estrela deve também possuir um raio de até 2 vezes o raio do Sol.

12 exoplanetas foram encontrados com base no período orbital proxímo ao da Terra, porém todos eles sendo 
maiores que a Terra e possivelmente com suas respectivas massas também maiores (A massa não foi utilizada nessa análise pela baixa quantidade
de dados disponível neste campo na base de dados);

75% dos exoplanetas encontrados possuem ao menos 3 vezes o raio da Terra e desses 75%, temos 44% deles sendo no minímo 6 vezes maior que o nosso
planeta, portanto eles devem ser planetas gasosos;

Outra interessante observação que podemos realizar é que entre 2014 e 2016 houve uma explosão na descoberta de exoplanetas, representando mais de 
51% dos exoplanetas encontrados de acordo com as métricas aplicadas na filtragem dos dados utilizados no estudo